src/Query.tsx

```diff
@@ -147,86 +184,92 @@ export default class Query<TData = any, TVariables = OperationVariables> extends
+      props,
+    } as any;
   }
 
   // For server-side rendering (see getDataFromTree.ts)
   fetchData(): Promise<ApolloQueryResult<any>> | boolean {
     if (this.props.skip) return false;
     // pull off react options
-    const { children, ssr, displayName, skip, client, ...opts } = this.props;
+    const { children, ssr, displayName, skip, client, asyncMode, ...opts } = this.props;
 
     let { fetchPolicy } = opts;
     if (ssr === false) return false;
     if (fetchPolicy === 'network-only' || fetchPolicy === 'cache-and-network') {
       fetchPolicy = 'cache-first'; // ignore force fetch in SSR;
     }
 
-    const observable = this.client.watchQuery({
+    const observable = this.props.client.watchQuery({
       ...opts,
       fetchPolicy,
     });
-    const result = this.queryObservable!.currentResult();
+    const result = this.state.queryObservable!.currentResult();
 
     return result.loading ? observable.result() : false;
   }
 
   componentDidMount() {
     this.hasMounted = true;
     if (this.props.skip) return;
-    this.startQuerySubscription();
+    this.startQuerySubscription(this.state.queryObservable);
     if (this.refetcherQueue) {
       const { args, resolve, reject } = this.refetcherQueue;
-      this.queryObservable!
-        .refetch(args)
+      this.state
+        .queryObservable!.refetch(args)
         .then(resolve)
         .catch(reject);
     }
   }
 
-  componentWillReceiveProps(nextProps: QueryProps<TData, TVariables>, nextContext: QueryContext) {
+  componentDidUpdate(prevProps: InnerQueryProps<TData, TVariables>) {
     // the next render wants to skip
-    if (nextProps.skip && !this.props.skip) {
+    if (this.props.skip && !prevProps.skip) {
       this.removeQuerySubscription();
       return;
     }
 
-    const { client } = nextProps;
-    if (
-      shallowEqual(this.props, nextProps) &&
-      (this.client === client || this.client === nextContext.client)
-    ) {
-      return;
-    }
+    // if there are no changes to the props, don't do anything state wise
+    if (shallowEqual(this.props, prevProps)) return null;
 
-    if (this.client !== client && this.client !== nextContext.client) {
-      if (client) {
-        this.client = client;
-      } else {
-        this.client = nextContext.client;
-      }
-      this.removeQuerySubscription();
-      this.queryObservable = null;
+    // if the client or the actual operation changed, we need to clean up the subscription
+    if (this.props.client !== prevProps.client || this.props.query !== prevProps.query) {
       this.previousData = {};
-
-      this.updateQuery(nextProps);
-    }
-    if (this.props.query !== nextProps.query) {
       this.removeQuerySubscription();
     }
 
-    this.updateQuery(nextProps);
-    if (nextProps.skip) return;
-    this.startQuerySubscription();
+    if (this.props.skip) return;
+    // start a new subscription if we don't have one already
+    this.startQuerySubscription(this.state.queryObservable);
```

Comment:
Nit: in the interest of triggering each side-effect off of a specific, easy-to-understand change in props, I'd rewrite this as:

```js
if (!this.props.skip) {
  this.startQuerySubscription(this.state.queryObservable);
}
```

That way, you don't have to think about the accumulation of prior conditions from which we may have hit an early `return`.